File: app/schemas/leaderboard.py
"""
Leaderboard Response Schemas
Pydantic models for leaderboard and ranking endpoints
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List
from uuid import UUID
from pydantic import BaseModel, Field, ConfigDict


class BoxMetricsSummary(BaseModel):
    """Summary metrics for leaderboard view"""
    
    floor_price_usd: Optional[Decimal] = None
    floor_price_1d_change_pct: Optional[Decimal] = None
    daily_volume_usd: Optional[Decimal] = None
    unified_volume_7d_ema: Optional[Decimal] = None
    units_sold_count: Optional[int] = None
    active_listings_count: Optional[int] = None
    listed_percentage: Optional[Decimal] = None
    estimated_total_supply: Optional[int] = None
    liquidity_score: Optional[Decimal] = None
    days_to_20pct_increase: Optional[Decimal] = None
    expected_days_to_sell: Optional[Decimal] = None
    price_sparkline_1d: Optional[List[dict]] = None  # Array of {timestamp, price}
    
    model_config = ConfigDict(from_attributes=True)


class LeaderboardBoxResponse(BaseModel):
    """Individual box in leaderboard response"""
    
    id: UUID
    rank: int = Field(..., ge=1, description="Current rank position")
    rank_change_direction: Optional[str] = Field(None, pattern="^(up|down|same)$")
    rank_change_amount: Optional[int] = None
    product_name: str
    set_name: Optional[str] = None
    game_type: Optional[str] = None
    image_url: Optional[str] = None
    metrics: BoxMetricsSummary
    reprint_risk: str = Field(default="LOW", pattern="^(LOW|MEDIUM|HIGH)$")
    metric_date: date
    
    model_config = ConfigDict(from_attributes=True)


class ResponseMeta(BaseModel):
    """Metadata for paginated responses"""
    
    total: int = Field(..., ge=0)
    sort: str = "unified_volume_7d_ema"
    sort_direction: str = Field(default="desc", pattern="^(asc|desc)$")
    date: Optional[date] = None
    limit: int = Field(default=50, ge=1, le=100)
    offset: int = Field(default=0, ge=0)


class LeaderboardResponse(BaseModel):
    """Full leaderboard response"""
    
    data: List[LeaderboardBoxResponse]
    meta: ResponseMeta


class SparklineDataPoint(BaseModel):
    """Single sparkline data point"""
    
    timestamp: datetime
    price: Decimal
    
    model_config = ConfigDict(from_attributes=True)

File: frontend/lib/api/leaderboard.ts
/**
 * Leaderboard API
 * Functions for fetching leaderboard data
 */

import { getAuthToken } from '@/lib/api/client';

export interface LeaderboardParams {
  sort?: string;
  limit?: number;
  offset?: number;
  date?: string;
}

export interface BoxMetricsSummary {
  floor_price_usd?: number | null;
  floor_price_1d_change_pct?: number | null;
  floor_price_30d_change_pct?: number | null;
  daily_volume_usd?: number | null;
  unified_volume_7d_ema?: number | null;
  unified_volume_usd?: number | null; // 30d volume
  volume_7d?: number | null; // 7d rolling sum
  volume_30d?: number | null; // 30d rolling sum
  volume_1d_change_pct?: number | null; // day-over-day volume % change
  volume_7d_change_pct?: number | null; // week-over-week volume % change
  volume_30d_change_pct?: number | null; // month-over-month volume % change
  boxes_sold_per_day?: number | null;
  units_sold_count?: number | null;
  boxes_sold_30d_avg?: number | null;
  active_listings_count?: number | null;
  top_10_value_usd?: number | null;
  liquidity_score?: number | null;
  days_to_20pct_increase?: number | null;
  expected_days_to_sell?: number | null;
  boxes_added_today?: number | null;
  // eBay marketplace data (from Phase 1b — 130point.com scraper)
  ebay_sold_today?: number | null;
  ebay_daily_volume_usd?: number | null;
  ebay_median_price_usd?: number | null;
  ebay_active_listings?: number | null;
  ebay_active_low_price?: number | null;
  ebay_volume_30d_usd?: number | null;
  ebay_boxes_added_today?: number | null;
  ebay_boxes_removed_today?: number | null;
  combined_boxes_sold_today?: number | null;
  daily_volume_tcg_usd?: number | null;
  daily_volume_ebay_usd?: number | null;
}

export interface LeaderboardBox {
  id: string;
  rank: number;
  rank_change_direction?: 'up' | 'down' | 'same' | null;
  rank_change_amount?: number | null;
  product_name: string;
  set_name?: string | null;
  game_type?: string | null;
  image_url?: string | null;
  metrics: BoxMetricsSummary;
  reprint_risk: string;
  metric_date?: string | null;
}

export interface ResponseMeta {
  total: number;
  sort: string;
  sort_direction: string;
  date?: string | null;
  limit: number;
  offset: number;
}

export interface LeaderboardResponse {
  data: LeaderboardBox[];
  meta: ResponseMeta;
}

/**
 * Get leaderboard data
 * Uses Next.js API proxy route to avoid CORS issues
 */
export async function getLeaderboard(params: LeaderboardParams = {}): Promise<LeaderboardResponse> {
  const searchParams = new URLSearchParams();
  if (params.sort) searchParams.append('sort', params.sort);
  if (params.limit) searchParams.append('limit', params.limit.toString());
  if (params.offset) searchParams.append('offset', params.offset.toString());
  if (params.date) searchParams.append('date', params.date);

  // Use Next.js API proxy route (avoids CORS issues)
  const url = `/api/booster-boxes?${searchParams.toString()}`;

  // Get auth token for authentication
  const token = getAuthToken();
  
  // Build headers
  const headers: HeadersInit = {
    'Content-Type': 'application/json',
  };
  
  // Add authorization header if token exists
  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }

  console.log('Fetching leaderboard from proxy:', url);
  
  const response = await fetch(url, {
    method: 'GET',
    headers,
    cache: 'no-store',
  });

  console.log('Response status:', response.status, response.statusText);

  if (!response.ok) {
    const errorText = await response.text();
    console.error('API Error Response:', errorText);
    let errorData;
    try {
      errorData = JSON.parse(errorText);
    } catch {
      errorData = { detail: `HTTP ${response.status}: ${response.statusText}` };
    }
    
    // Handle 401 Unauthorized - clear token and redirect to login
    if (response.status === 401) {
      // Clear invalid/expired token
      const { removeAuthToken } = await import('@/lib/api/client');
      removeAuthToken();
      
      // Redirect to landing page (where user can login)
      if (typeof window !== 'undefined') {
        window.location.href = '/landing';
      }
      
      throw new Error('Authentication required. Please log in to continue.');
    }
    
    throw new Error(errorData.detail || `Failed to fetch leaderboard: ${response.status} ${response.statusText}`);
  }

  const data = await response.json();
  console.log('Leaderboard data received:', data);
  return data;
}

